### PR TITLE
CLI improvement #303

### PR DESCRIPTION
**Pull request by bukharim96**
This update adds the `init` command to the CLI. But, it pretty much does plenty useful things so far:
  - creates test suite structure
  - installs dependencies
  - creates starter template
  - sets up the jest reporter

More features could definitely be added. But I'd like to do that gradually.

It could be used as follows: `unmock init`.

`init [options] [dirname]` Sets up a new test suite
  - `--installer [installer]`  Specify Package Manager (yarn/npm)
  - `--offline` Specify whether to allow install from cache
  - `--verbose` Specify verbosity of output messages

[Read more](https://github.com/bukharim96/unmock-js/tree/dev/packages/unmock-cli#unmock-cli).

### CODE REVIEW

File: packages/unmock-cli/README.md [hunk 1]
```diff
@@ -9,4 +9,19 @@
...
+This package contains the base functions needed to build or extend the [unmock](../unmock/README.md) CLI.  It also contains the cli itself for testing purposes.
+
+## Usage: unmock `<command>`
+
+### Options:
+  - `-h, --help` output usage information
+
+### Commands:
+  - `init [options] [dirname]` Sets up a new test suite
+    - `--installer [installer]`  Specify Package Manager (yarn/npm)
```
Comment (ksaaskil) [inline packages/unmock-cli/README.md]: In future, maybe the CLI could also check existing `package-lock.json` or `yarn.lock` to automatically determine the manager?

File: packages/unmock-cli/src/commands/curl.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  const { data } = await axios({
+    ...(options.data ? { data: JSON.parse(options.data) } : {}),
+    ...(options.headers ? { headers: makeHeaders(options.headers) } : {}),
+    method: options.request || "GET",
+    url,
+  });
+  return data;
+};
+
+export const curl = async (url: string, options: any) => {
```
Comment (ksaaskil) [inline packages/unmock-cli/src/commands/curl.ts]: We probably don't want to do network calls from the CLI `curl` anymore, maybe the internal `curl` could use `unmock-fetch` to mock network calls with in-memory faking. But this is fine now as is!

File: packages/unmock-cli/src/commands/init.starter.tmp.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      users: [],
+      error: e
+    }
+  }
+};
+
+// Basic unmock test
+const { u } = unmock;
+
+signs = ['Aries', 'Taurus', 'Gemini', 'Cancer', 'Leo', 'Virgo', 'Libra', 'Scorpio', 'Sagittarius', 'Capricorn', 'Aquarius', 'Pisces'];
```
Comment (ksaaskil) [inline packages/unmock-cli/src/commands/init.starter.tmp.ts]: `const`?

File: packages/unmock-cli/src/commands/init.starter.tmp.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
+export default `\
+/**
+ * Unmock Starter Template
+ *
+ * This is just to get you started, feel free to delete this.
+ * Place your tests in this file.
+ */
+const unmock = require("unmock");
```
Comment (ksaaskil) [inline packages/unmock-cli/src/commands/init.starter.tmp.ts]: In future, we could also support creating TS templates with ES6 imports etc.

File: packages/unmock-cli/src/commands/list.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    .slice(0, -1)
+    .join(" ");
+
+const SNAP_SUFFIX = ".snap";
+const DEFAULT_SNAPSHOTS_FOLDER = ".snapshots";
+interface IStringMap {
+  [key: string]: string[];
+}
+
+export const listInternal = (logger: ILogger): Promise<undefined> => {
```
Comment (ksaaskil) [inline packages/unmock-cli/src/commands/list.ts]: Not sure how relevant this is anymore...

File: packages/unmock-cli/src/commands/open.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import * as child_process from "child_process";
+import * as glob from "glob";
+import WinstonLogger from "../utils/logger";
+
+export const open = (hash: string, options: any) => {
```
Comment (ksaaskil) [inline packages/unmock-cli/src/commands/open.ts]: Also this is not relevant anymore, I think.

File: packages/unmock-cli/src/index.ts [hunk 7]
```diff
@@ -1,7 +1,47 @@
...
 export default () => {
+  const collect = (val: string, memo: string[]) => {
+    memo.push(val);
+    return memo;
+  };
+
   program.usage("unmock <command>");
-  program.command("init");
+
+  program
```
Comment (ksaaskil) [inline packages/unmock-cli/src/index.ts]: Not for this PR, but I have very much enjoyed using [oclif](https://github.com/oclif/oclif) from Heroku, it makes it easy to create modular CLIs with easy testability.

**Comment by bukharim96** [comment]
Thanks @ksaaskil. I wasn't sure of the relativity of the previous commands either. Nevertheless, I just preserved them. I haven't used `oclif` before, it looks great, and I'd be willing to use it next time round.

File: packages/unmock-cli/README.md (hunk 1 above)
Comment (bukharim96) [reply packages/unmock-cli/README.md]: I just added a fix: [bd97f36](https://github.com/unmock/unmock-js/pull/327/commits/bd97f36dac08f7f767ca78c22d5d9aa31e8165b2).

File: packages/unmock-cli/src/commands/curl.ts (hunk 2 above)
Comment (bukharim96) [reply packages/unmock-cli/src/commands/curl.ts]: OK

File: packages/unmock-cli/src/commands/init.starter.tmp.ts (hunk 3 above)
Comment (bukharim96) [reply packages/unmock-cli/src/commands/init.starter.tmp.ts]: I wasn't sure what would be a great starter template, so I just pulled that of the site.

File: packages/unmock-cli/src/commands/init.starter.tmp.ts (hunk 4 above)
Comment (bukharim96) [reply packages/unmock-cli/src/commands/init.starter.tmp.ts]: Perhaps we could have a flag for it: `unmock init --useTs`

File: packages/unmock-cli/src/commands/list.ts (hunk 5 above)
Comment (bukharim96) [reply packages/unmock-cli/src/commands/list.ts]: That makes two of us.

File: packages/unmock-cli/src/index.ts (hunk 7 above)
Comment (bukharim96) [reply packages/unmock-cli/src/index.ts]: I'll note that.

**Comment by mikesol** [comment]
Great work! Let's try to get this documented fast before the knowledge gets stale.

**Pull request merged**